### DISCUSSION TITLE
Should Show a hint in script-editor for missing call super counterpart callback function (if really existed)

### DISCUSSION THREAD

**Discussion by jzq740176597** [godotengine/godot-proposals#6639] [Scripting]
### Describe the project you are working on

3d

### Describe the problem or limitation you are having in your project

**very easy to forgot call the super callback Make debug very difficult**

> _func _ready() -> void:
> ___#super._ready()_
> 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

show hint or warning (gd already has same feature) if missing call super counterpart callback function (if really existed)

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

make debug more easy & less error-prone especially change class hierarchy later

### If this enhancement will not be used often, can it be worked around with a few lines of script?

no

### Is there a reason why this should be core and not an add-on in the asset library?

core

**Comment by theraot** [comment]
I noticed Godot is not auto completing function overrides※. I would want a `@virtual` annotation※※, if only so the editor know to autocomplete these. And when doing so, it could add the super call right away※※※.

※: Of method declared in GDScript by me. Not native methods.

※※: On the virtual method on the base base. Yes, in GDScript, of course. Not native methods.

※※※: On both native virtual methods, and annotated GDScript methods.

**Reply by dalexeev** [reply]
> I noticed Godot is not auto completing function overrides. I would want a `@virtual` annotation, if only so the editor know to autocomplete these. And when doing so, it could add the super call right away.

This is a bug. For **script** methods, we need to autocomplete overrides. Native non-virtual methods cannot be overridden (only shadowed, which is treated as an error by default), so no completion is needed for them.

**Reply by theraot** [reply]
@dalexeev 

First, I'm talking about GDScript. There is no `@virtual` annotation. But yes, they are overriden, not shadowed (you can do the experiment※). Anyway, the way I see it, it is feature missing, not a bug on an existing feature.

※: Declare a base class with a custom method, then a derived class with a matching method, instantiate the derived class, then cast it as the base class and see which method gets called. If it was just shadowed you get the method in the base class, if it was overriden you get the method on the derived class.

Second, there are are virtual native methods, which you can override, and the editor auto completes them. They are marked virtual in the documentation. There is even an [Overridable functions](https://docs.godotengine.org/en/stable/tutorials/scripting/overridable_functions.html#overridable-functions) article. I bet you use them all the time. I'm not talking about those.

**Reply by theraot** [reply]
After reading what I just wrote, the missing feature is `@final`, sadly. All GDScript methods are virtual by default.

**Reply by dalexeev** [reply]
> First, I'm talking about GDScript. There is no `@virtual` annotation. But yes, they are overriden, not shadowed (you can do the experiment※)

See godotengine/godot#72608. Native non-virtual methods cannot be overridden (the engine will not call them in general). Native virtual methods and script methods can be overridden.

**Reply by theraot** [reply]
@dalexeev Correct. I also saw the edit. Still, as I said, I wasn't talking about non-virtual methods. I wanted it to autocomplete overriding a virtual method made in GDScript, for which I would have to tell Godot which ones are those, by using a `@virtual` annotation. Please notice I didn't say an `@overrides` annotation. I wanted to put `@virtual` on the GDScript method I want Godot to help me override. Makes sense? Doesn't matter, I'll edit my first post (edit: done).

**Reply by dalexeev** [reply]
I am currently working on the related code (see godotengine/godot#75691), I will try to add the missing code to the script method overrides.

As for accidentally forgetting to add `super()` when overriding a method: we could generate not just an empty body (`pass` or nothing), but also insert a commented out `super(params)`. See also godotengine/godot#75013.

This won't save you if you later want to override a method in the base class and forget about descendant classes, but in most scenarios it will be enough I think. At the very least, you'll see `super()` more often when overriding a method, making it less likely to forget it.

**Comment by dalexeev** [comment]
See also:

* #3936
* #4461
